File: web/test/pipeline-extraction.test.ts
import { describe, it, expect } from 'vitest';
import { NormalizedDocument, comment } from '../src/core/model/NormalizedDocument.js';
import { CompressionLevel } from '../src/core/model/CompressionLevel.js';
import { CompressionPipeline } from '../src/core/pipeline/CompressionPipeline.js';
import { CompressedContext } from '../src/core/model/CompressedContext.js';

function run(doc: NormalizedDocument): CompressedContext {
  return CompressionPipeline.standard().process(doc, CompressionLevel.FULL);
}

describe('CompressionPipeline extraction', () => {
  it('extracts decisions including rejections', () => {
    const doc = new NormalizedDocument();
    doc.key = 'ABC-1';
    doc.title = 'Pick a cache';
    doc.description =
      'We decided to use Redis for the session cache. Memcached was rejected because it lacks persistence.';

    const result = run(doc);
    expect(result.decisions.some((d) => d.includes('Redis'))).toBe(true);
    expect(result.decisions.some((d) => d.includes('Memcached'))).toBe(true);
  });

  it('extracts acceptance criteria from section and Gherkin', () => {
    const doc = new NormalizedDocument();
    doc.key = 'ABC-2';
    doc.title = 'Login';
    doc.description = [
      'Acceptance Criteria',
      '- User can log in with email and password',
      '- Account locks after 5 failed attempts',
      '',
      'Given a locked account',
      'When the user resets the password',
      'Then the account unlocks',
    ].join('\n');

    const result = run(doc);
    expect(result.acceptanceCriteria.length).toBeGreaterThanOrEqual(3);
    expect(result.acceptanceCriteria.some((a) => a.startsWith('Given'))).toBe(true);
  });

  it('does not mistake prose starting with "When" for a Gherkin scenario', () => {
    const doc = new NormalizedDocument();
    doc.key = 'ABC-5';
    doc.title = 'Booking';
    doc.description =
      'When a patient books a visit, the appointment must appear in the EHR so that staff stop re-typing it.';

    const result = run(doc);
    expect(result.acceptanceCriteria).toEqual([]);
  });

  it('collapses repeated comments into one decision', () => {
    const doc = new NormalizedDocument();
    doc.key = 'ABC-3';
    doc.title = 'Broker choice';
    doc.description = 'Pick the message broker for order events.';
    doc.comments.push(comment('a', '', "Let's use Kafka for order events.", false));
    doc.comments.push(comment('b', '', "Let's use Kafka for order events.", false));
    doc.comments.push(comment('c', '', 'lets use kafka for order events', false));

    const result = run(doc);
    const kafka = result.decisions.filter((d) => d.toLowerCase().includes('kafka'));
    expect(kafka.length).toBe(1);
    expect(result.ignoredContent.some((i) => i.includes('duplicate'))).toBe(true);
  });

  it('keeps only high-value content at the tiny level', () => {
    const doc = new NormalizedDocument();
    doc.key = 'ABC-4';
    doc.title = 'Notifications';
    doc.description = 'We decided to use Kafka for events.';
    doc.comments.push(
      comment(
        'a',
        '',
        'I was on vacation last week, sorry for the slow reply. The weather was great by the way.',
        false,
      ),
    );

    const full = CompressionPipeline.standard().process(doc, CompressionLevel.FULL);
    const tiny = CompressionPipeline.standard().process(doc, CompressionLevel.TINY);
    expect(tiny.compressionRatio).toBeGreaterThanOrEqual(full.compressionRatio);
    expect(tiny.decisions.some((d) => d.includes('Kafka'))).toBe(true);
  });
});

File: crunchy-core/src/test/java/io/crunchy/core/PipelineExtractionTest.java
package io.crunchy.core;

import io.crunchy.core.model.CompressedContext;
import io.crunchy.core.model.CompressionLevel;
import io.crunchy.core.model.NormalizedDocument;
import io.crunchy.core.pipeline.CompressionPipeline;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

class PipelineExtractionTest {

    private CompressedContext run(NormalizedDocument doc) {
        return CompressionPipeline.standard().process(doc, CompressionLevel.FULL);
    }

    @Test
    void extractsDecisionsIncludingRejections() {
        NormalizedDocument doc = new NormalizedDocument();
        doc.setKey("ABC-1");
        doc.setTitle("Pick a cache");
        doc.setDescription("We decided to use Redis for the session cache. Memcached was rejected because it lacks persistence.");

        CompressedContext result = run(doc);
        assertTrue(result.getDecisions().stream().anyMatch(d -> d.contains("Redis")),
                "expected a Redis decision, got: " + result.getDecisions());
        assertTrue(result.getDecisions().stream().anyMatch(d -> d.contains("Memcached")),
                "expected the Memcached rejection, got: " + result.getDecisions());
    }

    @Test
    void extractsAcceptanceCriteriaFromSectionAndGherkin() {
        NormalizedDocument doc = new NormalizedDocument();
        doc.setKey("ABC-2");
        doc.setTitle("Login");
        doc.setDescription("""
                Acceptance Criteria
                - User can log in with email and password
                - Account locks after 5 failed attempts

                Given a locked account
                When the user resets the password
                Then the account unlocks
                """);

        CompressedContext result = run(doc);
        assertTrue(result.getAcceptanceCriteria().size() >= 3,
                "expected section items plus a Gherkin scenario, got: " + result.getAcceptanceCriteria());
        assertTrue(result.getAcceptanceCriteria().stream().anyMatch(a -> a.startsWith("Given")),
                "Gherkin scenario missing: " + result.getAcceptanceCriteria());
    }

    @Test
    void proseStartingWithWhenIsNotAGherkinScenario() {
        NormalizedDocument doc = new NormalizedDocument();
        doc.setKey("ABC-5");
        doc.setTitle("Booking");
        doc.setDescription("When a patient books a visit, the appointment must appear in the EHR "
                + "so that staff stop re-typing it.");

        CompressedContext result = run(doc);
        assertTrue(result.getAcceptanceCriteria().isEmpty(),
                "single 'When ...' prose sentence must not become an AC: "
                        + result.getAcceptanceCriteria());
    }

    @Test
    void collapsesRepeatedComments() {
        NormalizedDocument doc = new NormalizedDocument();
        doc.setKey("ABC-3");
        doc.setTitle("Broker choice");
        doc.setDescription("Pick the message broker for order events.");
        doc.getComments().add(new NormalizedDocument.Comment("a", "", "Let's use Kafka for order events.", false));
        doc.getComments().add(new NormalizedDocument.Comment("b", "", "Let's use Kafka for order events.", false));
        doc.getComments().add(new NormalizedDocument.Comment("c", "", "lets use kafka for order events", false));

        CompressedContext result = run(doc);
        long kafkaDecisions = result.getDecisions().stream().filter(d -> d.toLowerCase().contains("kafka")).count();
        assertEquals(1, kafkaDecisions, "repeats must collapse to one decision: " + result.getDecisions());
        assertTrue(result.getIgnoredContent().stream().anyMatch(i -> i.contains("duplicate")),
                "ignoredContent should report collapsed duplicates: " + result.getIgnoredContent());
    }

    @Test
    void tinyLevelKeepsOnlyHighValueContent() {
        NormalizedDocument doc = new NormalizedDocument();
        doc.setKey("ABC-4");
        doc.setTitle("Notifications");
        doc.setDescription("We decided to use Kafka for events.");
        doc.getComments().add(new NormalizedDocument.Comment("a", "",
                "I was on vacation last week, sorry for the slow reply. The weather was great by the way.", false));

        CompressedContext full = CompressionPipeline.standard().process(doc, CompressionLevel.FULL);
        CompressedContext tiny = CompressionPipeline.standard().process(doc, CompressionLevel.TINY);
        assertTrue(tiny.getCompressionRatio() >= full.getCompressionRatio(),
                "tiny must compress at least as hard as full");
        assertTrue(tiny.getDecisions().stream().anyMatch(d -> d.contains("Kafka")),
                "decisions must survive tiny compression");
    }
}
